lib/membrane_wav/parser.ex

```diff
@@ -0,0 +1,243 @@
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :always,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(state) do
+    state = Map.merge(state, %{stage: :init, queue: <<>>})
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_start_of_stream(:input, _context, state) do
+    demand = {:input, @init_stage_size}
+
+    {{:ok, demand: demand}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _context, %{stage: :data} = state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  def handle_demand(
+        :output,
+        buffers_count,
+        :buffers,
+        _context,
+        %{stage: :data, frames_per_buffer: frames, caps: caps} = state
+      ) do
+    size = buffers_count * Caps.frames_to_bytes(frames, caps)
+
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  def handle_demand(:output, _size, _unit, _context, state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_event(_pad, event, _context, state) do
+    Membrane.Logger.debug("Received event #{inspect(event)}")
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_process(:input, buffer, context, %{stage: :data} = state) do
+    if context.pads.input.demand == 0 do
+      {{:ok, buffer: {:output, buffer}, redemand: :output}, state}
+    else
+      {{:ok, buffer: {:output, buffer}}, state}
+    end
+  end
+
+  def handle_process(:input, %Buffer{payload: payload} = _buffer, context, state)
+      when byte_size(payload) < context.pads.input.demand do
+    demand_fun = &(&1 - byte_size(payload))
+    state = Map.update(state, :queue, payload, &(&1 <> payload))
+
+    {{:ok, demand: {:input, demand_fun}}, state}
+  end
+
+  def handle_process(
+        :input,
+        %Buffer{payload: payload} = _buffer,
+        _context,
+        %{stage: :init, queue: queue} = state
+      ) do
+    "RIFF" <>
+      <<_file_size::binary-size(4)>> <>
+      "WAVE" <>
+      "fmt " <>
+      <<format_size::binary-size(4)>> <>
+      <<format::binary-size(2)>> = queue <> payload
+
+    check_format(format, format_size)
+
+    demand = {:input, @format_stage_size}
+    state = %{state | stage: :format, queue: <<>>}
+
+    {{:ok, demand: demand}, state}
+  end
+
+  def handle_process(
+        :input,
+        %Buffer{payload: payload} = _buffer,
+        _context,
+        %{stage: :format, queue: queue} = state
+      ) do
+    <<channels::binary-size(2)>> <>
+      <<sample_rate::binary-size(4)>> <>
+      <<_data_transmission_rate::binary-size(4)>> <>
+      <<_block_alignment_unit::binary-size(2)>> <>
+      <<bits_per_sample::binary-size(2)>> <>
+      <<next_header_element::binary-size(8)>> = queue <> payload
```

Comment:
I spoke with Mateusz, we looked at https://hexdocs.pm/elixir/Kernel.SpecialForms.html#%3C%3C%3E%3E/1 and decided we'd like to introduce a "rule" to use `bytes` for matching binaries, `bits` for bitstrings. To match that, for integers I'd use `::32-signed-little` and here you can pattern match all the numbers (`integer` and `unsigned` are default, you may keep `unsigned` if you want)
```suggestion
    <<channels::16-little, sample_rate:: 32-little, _data_transmission_rate::32, _block_alignment_unit:: 16, bits_per_sample::16-little, next_header_type::4-bytes, next_header_size::32-little>> = queue <> payload
```
Note that I've already split something you matched in the next `case`